# Задание 3. Задача с качеством уроков

In [1]:
import pandas as pd

# Читаем все файлы, которые будут необходимы для выполнения данного задания

lessons_data = pd.read_csv('/Users/garynich/Documents/Tetrica School/tech_quality/lessons.txt', sep='|')
lessons_data = lessons_data[1:] #Удаляем первую строку, так как она несет в себе мусор -  особенность чтения  txt
lessons_data = lessons_data[:-1]

participants_data = pd.read_csv('/Users/garynich/Documents/Tetrica School/tech_quality/participants.txt', sep='|')
participants_data = participants_data[1:] #Удаляем первую строку, так как она несет в себе мусор -  особенность чтения  txt
participants_data = participants_data[:-1]

quality_data = pd.read_csv('/Users/garynich/Documents/Tetrica School/tech_quality/quality.txt', sep='|')
quality_data = quality_data[1:] #Удаляем первую строку, так как она несет в себе мусор -  особенность чтения  txt
quality_data = quality_data[:-1]

users_data = pd.read_csv('/Users/garynich/Documents/Tetrica School/tech_quality/users.txt', sep='|')
users_data = users_data[1:] #Удаляем первую строку, так как она несет в себе мусор -  особенность чтения  txt
users_data = users_data[:-1]



In [2]:
lessons_data.columns = ['id', 'event_id', 'subject', 'scheduled_time']

lessons_data['event_id'] = lessons_data['event_id'].astype(int) #Переводим id в целочисленный тип для упрощения работы
lessons_data['scheduled_time'] =  pd.to_datetime(lessons_data['scheduled_time'])
lessons_data['mow_time'] = lessons_data['scheduled_time'] + pd.to_timedelta(3, unit='h') #Создаем столбец 
            #с проведением урока по московскому времени
lessons_data['subject'] = lessons_data['subject'].str.strip() # Удаляем пробелы со всех сторон у значения


for row in lessons_data.columns:
    print('Количество пропусков в столбце ', row, '=', lessons_data[row].isnull().sum())
print(lessons_data.info())

lessons_data.head(3)

Количество пропусков в столбце  id = 0
Количество пропусков в столбце  event_id = 0
Количество пропусков в столбце  subject = 0
Количество пропусков в столбце  scheduled_time = 0
Количество пропусков в столбце  mow_time = 0
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 378 entries, 1 to 378
Data columns (total 5 columns):
id                378 non-null object
event_id          378 non-null int64
subject           378 non-null object
scheduled_time    378 non-null datetime64[ns]
mow_time          378 non-null datetime64[ns]
dtypes: datetime64[ns](2), int64(1), object(2)
memory usage: 14.8+ KB
None


id  event_id subject  \
1   1e7bb408-cfef-4a9f-8328-351c9483a64c      38114    phys   
2   6d8e59d9-a7c8-4bb3-8ff3-99cd07acdf1a      51568      it   
3   62e1a078-33de-47c1-99d2-845b1daca56f      52790    hist   

       scheduled_time            mow_time  
1 2020-01-19 12:00:00 2020-01-19 15:00:00  
2 2020-01-19 13:00:00 2020-01-19 16:00:00  
3 2020-01-19 13:00:00 2020-01-19 16:00:00

In [3]:
participants_data.columns = ['event_id', 'user_id']
participants_data['user_id'] = participants_data['user_id'].astype(str)
participants_data['event_id'] = participants_data['event_id'].astype(int)
for row in participants_data.columns:
    print('Количество пропусков в столбце ', row, '=', participants_data[row].isnull().sum())
print(participants_data.info())


participants_data.head(3)

Количество пропусков в столбце  event_id = 0
Количество пропусков в столбце  user_id = 0
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 743 entries, 1 to 743
Data columns (total 2 columns):
event_id    743 non-null int64
user_id     743 non-null object
dtypes: int64(1), object(1)
memory usage: 11.7+ KB
None


event_id                                user_id
1     38114   e28351f5-4ccb-4549-8647-d43f2b15e7b8
2     38114   4df2832a-1d63-4453-9659-43993fc35996
3     51568   bb1c0bc8-1212-452b-97a0-439d4a2169e2

In [4]:
quality_data.columns = ['lesson_id', 'tech_quality']
quality_data['tech_quality'] = quality_data['tech_quality'].str.strip() # Удаляем пробелы со всех сторон у значения
quality_data['tech_quality'] = pd.to_numeric(quality_data['tech_quality'])

for row in quality_data.columns:
    print('Количество пропусков в столбце ', row, '=', quality_data[row].isnull().sum())
print(quality_data.info())

quality_data.head(3)

Количество пропусков в столбце  lesson_id = 0
Количество пропусков в столбце  tech_quality = 16
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 365 entries, 1 to 365
Data columns (total 2 columns):
lesson_id       365 non-null object
tech_quality    349 non-null float64
dtypes: float64(1), object(1)
memory usage: 5.8+ KB
None


lesson_id  tech_quality
1   6d8e59d9-a7c8-4bb3-8ff3-99cd07acdf1a            5.0
2   62e1a078-33de-47c1-99d2-845b1daca56f            5.0
3   62e1a078-33de-47c1-99d2-845b1daca56f            5.0

In [5]:
users_data.columns = ['user_id', 'role']
users_data['user_id'] = users_data['user_id'].astype(str)
for row in users_data.columns:
    print('Количество пропусков в столбце ', row, '=', users_data[row].isnull().sum())
print(users_data.info())

users_data.head(3)

Количество пропусков в столбце  user_id = 0
Количество пропусков в столбце  role = 0
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 743 entries, 1 to 743
Data columns (total 2 columns):
user_id    743 non-null object
role       743 non-null object
dtypes: object(2)
memory usage: 11.7+ KB
None


user_id    role
1   e28351f5-4ccb-4549-8647-d43f2b15e7b8    pupil
2   4df2832a-1d63-4453-9659-43993fc35996    tutor
3   bb1c0bc8-1212-452b-97a0-439d4a2169e2    pupil

Выполнена предобработка предоставленных файлов. Данные переведены в тип, который пригоден для дальнейшей работы. 

##  1. Найти все уроки по физике (subject=phys).

In [6]:
lessons_data  = lessons_data.query('subject == "phys"')
lessons_data.head(5)

id  event_id subject  \
1   1e7bb408-cfef-4a9f-8328-351c9483a64c      38114    phys   
4   00fc6685-f53a-49bb-b960-5e0042fd3852      51341    phys   
5   4cadf623-82e6-422f-a342-acf978302fb2      55048    phys   
7   e597a79a-3f68-4d15-bc77-c7000bcf8e52      53026    phys   
9   ea6f0bf3-bcfc-4555-9b16-cc2cfe722e17      51645    phys   

       scheduled_time            mow_time  
1 2020-01-19 12:00:00 2020-01-19 15:00:00  
4 2020-01-17 12:00:00 2020-01-17 15:00:00  
5 2020-01-19 14:00:00 2020-01-19 17:00:00  
7 2020-01-19 15:00:00 2020-01-19 18:00:00  
9 2020-01-19 14:00:00 2020-01-19 17:00:00

## 2. В каждый день для каждого репетитора посчитать среднюю арифметическую оценку за его уроки (учитывать только уроки из п. 1).

In [7]:
rep = quality_data.pivot_table(index = 'lesson_id', values = 'tech_quality', aggfunc = ['count', 'sum'])
rep.columns = ['count_quality', 'sum_quality']
rep.reset_index(inplace = True)
rep.head(5)

lesson_id  count_quality  sum_quality
0   00fc6685-f53a-49bb-b960-5e0042fd3852               2         10.0
1   01f717d8-baaf-427d-a927-08c79713ae79               2          7.0
2   024d415a-4543-4e73-88c6-f4609e1bf1ed               2          9.0
3   03f4f4a8-c60c-41d7-a1cb-17b2d36ee183               1          5.0
4   056c3d91-dce7-4071-91d3-1de816ee63a3               1          4.0

In [8]:
lessons_datas = pd.merge(lessons_data, rep, left_on='id', right_on='lesson_id')

lessons_datas['mow_day'] = lessons_datas['mow_time'].dt.date
lessons_datas['mean_quality'] = lessons_datas['sum_quality'] /  lessons_datas['count_quality']
lessons_datas.drop(['scheduled_time', 'lesson_id', 'mow_time'], axis=1, inplace=True)


part_with_roles = participants_data.join(users_data, lsuffix='_l')
part_with_roles.drop(['user_id_l'], axis=1, inplace=True)

teachers_with_event = part_with_roles.query('role == " tutor"')

finish_data  = pd.merge(lessons_datas, teachers_with_event, on = 'event_id')
finish_data.head(5)

id  event_id subject  count_quality  \
0   00fc6685-f53a-49bb-b960-5e0042fd3852      51341    phys              2   
1   4cadf623-82e6-422f-a342-acf978302fb2      55048    phys              1   
2   e597a79a-3f68-4d15-bc77-c7000bcf8e52      53026    phys              1   
3   ea6f0bf3-bcfc-4555-9b16-cc2cfe722e17      51645    phys              1   
4   ea6f0bf3-bcfc-4555-9b16-cc2cfe722e17      51645    phys              1   

   sum_quality     mow_day  mean_quality  \
0         10.0  2020-01-17           5.0   
1          5.0  2020-01-19           5.0   
2          5.0  2020-01-19           5.0   
3          5.0  2020-01-19           5.0   
4          5.0  2020-01-19           5.0   

                                  user_id    role  
0   30a19496-bdaf-461c-abbc-2709ae520201    tutor  
1   43efce48-94b2-4412-857f-223d45969008    tutor  
2   43efce48-94b2-4412-857f-223d45969008    tutor  
3   8fe03f08-8581-430c-a590-9888ab36deb3    tutor  
4   8fe03f08-8581-430c-a590-9888ab36deb3    tutor

## 3. Найти учителя, который в этот день имеет самую низкую среднюю арифметическую оценку за уроки 

In [9]:
pre = finish_data.groupby(['mow_day', 'user_id']).agg({'mean_quality':'mean'}).round(2).reset_index()
report = pre.pivot_table(index = 'mow_day', aggfunc = 'min')

report[['user_id', 'mean_quality']].reset_index()


mow_day                                 user_id  mean_quality
0  2020-01-11   2fa2ab62-f1b0-4036-872f-bcfd9a8686ff           4.00
1  2020-01-12   1b99ff07-a520-44dc-9f1b-38d98c85fcb3           4.90
2  2020-01-13   2fa2ab62-f1b0-4036-872f-bcfd9a8686ff           5.00
3  2020-01-14   2fa2ab62-f1b0-4036-872f-bcfd9a8686ff           4.00
4  2020-01-15   1b99ff07-a520-44dc-9f1b-38d98c85fcb3           5.00
5  2020-01-16   696c838e-c054-4e9f-a51a-50bf5660f364           4.50
6  2020-01-17   30a19496-bdaf-461c-abbc-2709ae520201           4.50
7  2020-01-18   2fa2ab62-f1b0-4036-872f-bcfd9a8686ff           3.50
8  2020-01-19   1b99ff07-a520-44dc-9f1b-38d98c85fcb3           4.50
9  2020-01-20   2fa2ab62-f1b0-4036-872f-bcfd9a8686ff           4.38

# Ответ:

    Для каждого дня найден пользователь, который получил самую низкую среднюю оценку.

<table border="1" class="dataframe">\n  <thead>\n    <tr style="text-align: right;">\n      <th></th>\n      <th>mow_day</th>\n      <th>user_id</th>\n      <th>mean_quality</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>2020-01-11</td>\n      <td>2fa2ab62-f1b0-4036-872f-bcfd9a8686ff</td>\n      <td>4.00</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>2020-01-12</td>\n      <td>1b99ff07-a520-44dc-9f1b-38d98c85fcb3</td>\n      <td>4.90</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>2020-01-13</td>\n      <td>2fa2ab62-f1b0-4036-872f-bcfd9a8686ff</td>\n      <td>5.00</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>2020-01-14</td>\n      <td>2fa2ab62-f1b0-4036-872f-bcfd9a8686ff</td>\n      <td>4.00</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>2020-01-15</td>\n      <td>1b99ff07-a520-44dc-9f1b-38d98c85fcb3</td>\n      <td>5.00</td>\n    </tr>\n    <tr>\n      <th>5</th>\n      <td>2020-01-16</td>\n      <td>696c838e-c054-4e9f-a51a-50bf5660f364</td>\n      <td>4.50</td>\n    </tr>\n    <tr>\n      <th>6</th>\n      <td>2020-01-17</td>\n      <td>30a19496-bdaf-461c-abbc-2709ae520201</td>\n      <td>4.50</td>\n    </tr>\n    <tr>\n      <th>7</th>\n      <td>2020-01-18</td>\n      <td>2fa2ab62-f1b0-4036-872f-bcfd9a8686ff</td>\n      <td>3.50</td>\n    </tr>\n    <tr>\n      <th>8</th>\n      <td>2020-01-19</td>\n      <td>1b99ff07-a520-44dc-9f1b-38d98c85fcb3</td>\n      <td>4.50</td>\n    </tr>\n    <tr>\n      <th>9</th>\n      <td>2020-01-20</td>\n      <td>2fa2ab62-f1b0-4036-872f-bcfd9a8686ff</td>\n      <td>4.38</td>\n    </tr>\n  </tbody>\n</table>